### ISSUE TITLE
Do we really require wheelhouse passthrough?

### ISSUE THREAD

**Issue by armab** [StackStorm/st2-packages#23]
Looks like ghost bug.

From time to time I get this error in CircleCI `wheezy` build:
https://circleci.com/gh/StackStorm/st2-packages/36

```sh
[package: st2common]      dpkg-shlibdeps: error: couldn't find library libffi.so.6 needed by debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/_cffi_backend.so (ELF format: 'elf64-x86-64'; RPATH: '')
[package: st2common]      dpkg-shlibdeps: warning: package could avoid a useless dependency if debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/msgpack/_unpacker.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/msgpack/_packer.so debian/st2common/usr/share/python/st2common/bin/python were not linked against libgcc_s.so.1 (they use none of the library's symbols)
[package: st2common]      dpkg-shlibdeps: warning: package could avoid a useless dependency if debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/msgpack/_unpacker.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/msgpack/_packer.so were not linked against libstdc++.so.6 (they use none of the library's symbols)
[package: st2common]      dpkg-shlibdeps: error: cannot continue due to the error above
[package: st2common]      Note: libraries are not searched in other binary packages that do not have any shlibs or symbols file.
[package: st2common]      To help dpkg-shlibdeps find private libraries, you might need to set LD_LIBRARY_PATH.
[package: st2common]      dh_shlibdeps: dpkg-shlibdeps -Tdebian/st2common.substvars debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Hash/_MD2.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Hash/_MD4.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Hash/_SHA256.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Hash/_SHA224.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Hash/_SHA384.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Hash/_SHA512.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Hash/_RIPEMD160.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_AES.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_ARC2.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_Blowfish.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_CAST.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_DES.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_DES3.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_ARC4.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Cipher/_XOR.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Util/strxor.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/Crypto/Util/_counter.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/markupsafe/_speedups.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/bson/_cbson.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/pymongo/_cmessage.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/cryptography/hazmat/bindings/_constant_time.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/cryptography/hazmat/bindings/_padding.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/msgpack/_packer.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/msgpack/_unpacker.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/wrapt/_wrappers.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/greenlet.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/_yaml.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/_cffi_backend.so debian/st2common/usr/share/python/st2common/lib/python2.7/site-packages/netifaces.so debian/st2common/usr/share/python/st2common/bin/python returned exit code 2
[package: st2common]      make: *** [binary] Error 2

.circle/docker-compose.sh build ${DISTRO} returned exit code 1

[package: st2common] dpkg-buildpackage: error: debian/rules binary gave error exit status 2 Action failed: .circle/docker-compose.sh build ${DISTRO}
```

#### More info
Sometimes just pushing `Rebuild` button works, sometimes not: several retries needed.

**Comment by dennybaa** [comment]
It's not a ghost bug. I've already located this one, problem is the cache in parallel setup scenario.
Still don't know the solution, need to figure out cache save/restore algorithm.

**Comment by armab** [comment]
I got that error when running 1 CircleCI VM (without many CI parallelization).
Or you mean parallel packages builds inside one VM?

**Comment by dennybaa** [comment]
Clean cache and repeat with x1 parallelism.

**Comment by armab** [comment]
Had luck to find the better reason.

This bug is caused by `wheelhouse`: https://github.com/StackStorm/st2-packages/commit/b5add4a927078ce7312b8746a19b04e9a7da66b1

Though `~/.cache/pip` works well, see successful build with `~/wheelhouse` cache disabled: https://circleci.com/gh/StackStorm/st2-packages/96

----

Another thing I noted, after disabling wheelhouse caching, the build time wasn't changed, so probably `wheelhouse` needs more debugging and adjusting?

As https://wheel.readthedocs.org/en/latest/ say, the improvement by using `wheelhouse` should be dramatic, or it's very specific? This is the topic where I have zero knowledge.

**Comment by dennybaa** [comment]
Above I've told it's a caching problem, and it's so. Just because the wheelhouse cache is OS specific :)

**Comment by dennybaa** [comment]
About performance it saves about 2 minutes, that what my tests showed, not that much indeed...

**Comment by dennybaa** [comment]
Wheelhouse passthrough requires cache collection from all build containers and individual restoration, due to wheels being specific to OS Distro. This kind of setup might be complicated.

* We need to measure the goal of kind of cache.
* Decide whether to use it or not.

**Comment by armab** [comment]
Are you saying that wheelhouse cache should be distro-based?

If so, we can do this way:
```yaml
dependencies:
  cache_directories:
    - ~/.cache/pip
    - ~/wheelhouse/wheezy
    - ~/wheelhouse/jessie
    - ~/wheehouse/trusty
    - ~/wheehouse/centos7
    - ~/wheehouse/centos6
```
Looks pretty nice.

But could you recognize which dir to use in your scripts? Eg. we need to inject ${DISTRO} var from CI?

**Comment by armab** [comment]
Related thing after recognizing `${DISTRO}` in build environment could be: 
```yaml
  artifacts:
    - /tmp/st2-packages/wheezy
    - /tmp/st2-packages/jessie
    - /tmp/st2-packages/trusty
    - /tmp/st2-packages/centos7
    - /tmp/st2-packages/centos6
```
So when running `docker-compose run wheezy` - packages are stored in `/tmp/st2-packages/wheezy` and so on for every distro.

That will simplify things later when we'll reach deployment stage.

**Comment by dennybaa** [comment]
It might look from the first glance. Unfortunately cache is saved only on container 0.  This is the same story with collecting artifacts.

**Comment by armab** [comment]
Are you sure? I see `Restore cache` and `Save cache` run for every CircleCI container.

**Comment by dennybaa** [comment]
yes. https://discuss.circleci.com/t/cache-save-restore-algorithm/759

**Comment by armab** [comment]
Thanks for insights.
That's pretty unexpected limitation from Circle :disappointed:

**Comment by dennybaa** [comment]
It's not that unexpected. I don't really see any solution to satisfy all 100% of clients needs. It designed to build that what 80% percent of clients build, and we are those lucky 20%)

In our case it would require separate cache handling, and it's not that anyone wants to store. You use parallelism x16, then you store x16 more :) that sucks.

**Comment by armab** [comment]
Haha, I'm always in "lucky" 20%.
Same with Drone, same with every other software. Something always is not enough :)

Not surprizing then why `wheelhouse` didn't work. If you're saying about `2 mins` speedup for `wheezy` build which takes 9 mins, then for more heavy build like `centos7` the speedup will be really huge.
By the reason of caching only `0` node I didn't see any speedup when running `x3` parallelism, that clarifies everything.

It feels like we should work on storing cache in `S3` in some future, assign it lower priority and return someday. Definitely good thing to have later :+1:

**Comment by dennybaa** [comment]
remove duplicate #14

**Comment by dennybaa** [comment]
obsolete
